### PR TITLE
Spruce up async handling in OnNavigateAsync callback in Blazor router

### PR DESCRIPTION
**Pull request by captainsafia**
This PR introduces some modifications to improve handling of async operations and cancellations in OnNavigateAsync. The changes included:

* Use `Cancel` instead of `Dispose` on the `_onNavigateCts` inside `OnNavigateAsync`. Dispose does not emit a cancellation and simple breaks down registrations on the token source.
* Remove the `_onNavigate?.Dispose` call in the `Dispose` method of the router. Disposing of the cancellation token source isn't meaningful in our context and introduces more headaches than benefit. The [Dispose method on the CTS](https://github.com/dotnet/runtime/blob/ea2b09beea535abf960651c59359e40f99f2fecf/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs#L407) disposes of timers, wait handles, and linked tokens on a CTS. Since we don't use any of these in our code, we don't need to dispose. This way, we don't have to worry about the unintended consequences of trying to cancel a disposed token.
* Adds the `TaskCreationOptions.RunContinuationsAsynchronously` constructor parameter to the `TaskCompletionSource` that observes the cancellation token. Without this option, the `cancellationTcs` continuation will run inline when triggered. When we visit Route A, then Route B, the cancellation of the CTS that occurs in the second invocation of `OnNavigateAsync` will continue running the rest of A's `OnNavigateAsync` and triggering a `Refresh`.
* Checks to see if a task was completed or cancelled to avoid extraneously rendering a route's component if the OnNavigate was cancelled.
* Checks to see if the completion or cancellation of the previous task was fully finalized before starting the next OnNavigateAsync task.
* Adds unit tests for `OnNavigateAsync` and changes a few modifiers on `Router` to support testing.

Addresses #23789

### CODE REVIEW

**Comment by mrpmorris** [comment]
Did you see my comment on the last PR about naming it OnNavigatedAsync?

I'm not looking for a response, just want to ensure it wasn't missed.

File: src/Components/Components/src/Routing/Router.cs
Comment (gfoidl) [inline src/Components/Components/src/Routing/Router.cs:26]:
```suggestion
        internal RenderHandle RenderHandle { get; set; }
```
Naming conventions when making this a property.

File: src/Components/Components/src/Routing/Router.cs
Comment (gfoidl) [inline src/Components/Components/src/Routing/Router.cs:203]:
```suggestion
                return true;
```

File: src/Components/Components/src/Routing/Router.cs
Comment (gfoidl) [inline src/Components/Components/src/Routing/Router.cs:233]:
```suggestion
            return task == completedTask;
```

**Comment by captainsafia** [comment]
> Did you see my comment on the last PR about naming it OnNavigatedAsync?

I did see the comment but haven't thought much about `OnNavigateAsync` vs `OnNavigatedAsync` yet.

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs:26]: Good call! I'll hold off on accepting the suggestion since it requires changing invocations in the Router class. Also, I'd like to see if there is a way to avoid having to make this change at all.

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [inline src/Components/Components/src/Routing/Router.cs:152]: For tests so we can mock out `Refresh`.

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [inline src/Components/Components/src/Routing/Router.cs:233]: For tests.

File: src/Components/Components/src/Routing/Router.cs
Comment (JamesNK) [inline src/Components/Components/src/Routing/Router.cs:217]: Because the previous OnNavigate is always awaited, am I right that there can only be one call in progress?

How much do you care about performance? You might be able to change the TaskCompletionSource here to use a task backed by https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.sources.manualresetvaluetasksourcecore-1?view=dotnet-plat-ext-3.1. That would basically allow for a resettable completion source.

Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs:217]: > Because the previous OnNavigate is always awaited, am I right that there can only be one call in progress?

Correct.

> You might be able to change the TaskCompletionSource here to use a task backed by https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.sources.manualresetvaluetasksourcecore-1?view=dotnet-plat-ext-3.1. That would basically allow for a resettable completion source.

Interesting. I wasn't aware of this API. Might be worthwhile to explore. Strikes me as a bit more readable to use here.

File: src/Components/Components/src/Routing/Router.cs
Comment (SteveSandersonMS) [reply src/Components/Components/src/Routing/Router.cs:152]: `internal virtual` is fine, but `protected internal virtual` makes it part of the public API surface. Can this just be `internal virtual`? As long as the test class has InternalsVisibleTo, it should be allowed to override it.

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs]: Worth doing it for consistency with elsewhere in our codebase.

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs:217]: I played around with this. I like not having to do the allocation for a new TCS at every invocation but the API for MRVTSC is a little meh in some places (e.g. no way to set task creation options in a constructor since its a struct).

For now, I'll avoid prematurely optimizing this and see if users run into perf issues when using this in practice. Thanks for the recommendation, tho!

Comment (javiercn) [reply src/Components/Components/src/Routing/Router.cs:217]: I wouldn't worry about an allocation in this case, seems like a premature optimization.

File: src/Components/Components/src/Routing/Router.cs
Comment (pranavkm) [inline src/Components/Components/src/Routing/Router.cs:195]:
```suggestion
        private async ValueTask<bool> RunOnNavigateAsync(string path, Task previousOnNavigate)
```

File: src/Components/Components/src/Routing/Router.cs
Comment (pranavkm) [inline src/Components/Components/src/Routing/Router.cs:201]: If I follow the code correctly, RefreshRouteTable is invoked in this case. Why would we need to do that?

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (pranavkm) [inline src/Components/Components/test/Routing/RouterTest.cs:35]: Why is it ok to not await these tasks?

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (pranavkm) [inline src/Components/Components/test/Routing/RouterTest.cs:82]:
```suggestion
            router.Verify(x => x.Refresh(false), Times.Once());
```

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs:201]: If this returns `true`, then we call `Refresh` which renders the route's component. `RefreshRouteTable` is called inside `Refresh` but will only change the route table if `AdditionalAssemblies` is modified.

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs]: Yeah, we can use the `InternalsVisibleTo` attribute here. Updated.

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (captainsafia) [reply src/Components/Components/test/Routing/RouterTest.cs:35]: Starting the task without awaiting better simulates how the invocation ends up being made when we call it from `OnLocationChanged` in the router. This presents an issue when the tests are running in CI though since we don't wait for the async tasks to complete before validating what they did...

File: src/Components/Components/src/Routing/Router.cs
Comment (halter73) [inline src/Components/Components/src/Routing/Router.cs:241]: Should we leave a note that the above logic would need a lock if used in a preemptive multitasking environment?

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (halter73) [reply src/Components/Components/test/Routing/RouterTest.cs:35]: Generally, when Tasks aren't directly awaited in a test, I still like storing the Task and asserting that it's incomplete. Later on in the test when I expect the Task to be completed (if ever), I then assert the Task is complete and verify it has the expected result (if any).

In this case, RunOnNavigateWithRefreshAsync should be returning a completed Task, right? If that's the case, I would just "await" it.

File: src/Components/Components/src/Routing/Router.cs
Comment (halter73) [inline src/Components/Components/src/Routing/Router.cs:205]: Nit:

```suggestion
            // cancel that existing CTS.
```

**Review by halter73 — APPROVED** [review]

**Review by pranavkm — APPROVED** [review]

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs]: I made this change in a follow-up commit. Thanks for recommending the clarification!

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs:243]: Yes, we discussed this during our meeting but I forgot to include that here. Thanks!

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (captainsafia) [reply src/Components/Components/test/Routing/RouterTest.cs]: Yes, `RunOnNavigateWithRefreshAsync` returns a completed task. Updated this to use `await`.

File: src/Components/Components/src/Routing/Router.cs
Comment (javiercn) [inline src/Components/Components/src/Routing/Router.cs:205]: In general, this comment adds no value. It simply states what the line below is already doing, I would suggest you update it with the reasons why it is important to know to use Cancel vs Dispose here. Essentially, explain the why instead of the what. Otherwise, simply remove the comment.

File: src/Components/Components/src/Routing/Router.cs
Comment (javiercn) [inline src/Components/Components/src/Routing/Router.cs:241]: I might be wrong here, but if we were to support multi-threading we would still be single threaded on the UI, since changing that would be a massive breaking change of epic proportions. We would likely switch to use a synchronization context in the same way we do on the server.

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (javiercn) [inline src/Components/Components/test/Routing/RouterTest.cs:60]: The order of these two parameters is reversed.

File: src/Components/Components/src/Routing/Router.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/Routing/Router.cs:209]: I'm sure this `await previousOnNavigate` is for a good reason, but it seems subtle. I can see that due to the chain of events triggered by L206, the previous task will immediately become cancelled, so the await on L209 shouldn't have to really wait.

So it makes me wonder why we couldn't just ignore `previousOnNavigate`, since we know it will have been cancelled and that its continuation is going to no-op (not refresh the UI).

Also I spent a while being confused about error handling here, because I thought that if `previousOnNavigate` had failed or been cancelled, we'd be blocking all subsequent navigation because we're awaiting it. Then later I realised that `previousOnNavigate` can never fail or be cancelled because it's entirely managed within `RunOnNavigateWithRefreshAsync`, which only ever sets it to completed.

So, it would be good to clarify in the comment why we want to await this, and why we know it will never fail or be cancelled :)

Comment (SteveSandersonMS) [reply src/Components/Components/src/Routing/Router.cs:209]: Update: in fact I realise now that `previousOnNavigate` will *not* actually have been cancelled; it will just have been completed due to the chain of events triggered by L206. So the existing comment is a bit misleading since it suggests it might be cancelled.

All these CTSs and TCSs are quite hard to keep track of! I do still wonder why we need to await `previousOnNavigate` here.

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (SteveSandersonMS) [inline src/Components/Components/test/Routing/RouterTest.cs:102]: These don't seem to be used.

File: src/Components/Components/src/Routing/Router.cs
Comment (SteveSandersonMS) [inline src/Components/Components/src/Routing/Router.cs:230]: I'm unclear about this logic.

 * If the developer's task is cancelled by them, is the intended behavior that we treat it the same as if it resolved as success? That is, we do perform the UI refresh. That appears to be the behavior here, and I suspect that's the best choice because we don't want to leave things in a "loading" state, and we don't have a way to truly cancel navigation and put the URL back to what it was before.
 * But, how do we know it was cancelled by them? Inside their `OnNavigateAsync` callback, they may well be using the cancellation token we gave them. So if that token fires, their returned task may be immediately cancelled as a result. If that happens it seems like we have a race condition: either `task` or `cancellationTcs.Task` might happen to resolve first, since they were both triggered by the same underlying `_onNavigateCts.Token`. It makes me wonder if L230 really should say `return !_onNavigateCts.Token.IsCancellationRequested` or something like that.
 * Also, if the developer's task resolves as faulted, then the logic here will disregard that fault and treat it the same as completing successfully. It seems like instead we should be throwing the exception upstream. I know that we don't yet have a way to truly pass it back into the renderer, but as it stands we're discarding the exception in two places (here on L230 and also on L264). Shouldn't we be throwing here so the fault goes to one place (L264) where we can do something useful with it later?

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (SteveSandersonMS) [inline src/Components/Components/test/Routing/RouterTest.cs:90]: It looks like you have well covered the cases where cancellation occurs due to a second navigation.

I'd love to see some additional test cases covering scenarios where the OnNavigateAsync task itself is what completes first, in success, failure, and cancelled status. For example:

- OnNavigateAsync task fails
  - While it's the activate navigation
  - After some other navigation has started
- OnNavigateAsync task is cancelled
  - While it's the activate navigation
  - After some other navigation has started

**Comment by SteveSandersonMS** [comment]
@captainsafia Sorry these comments are after you merged. It was totally reasonable to merge at the time you did. But it would be great if we could follow up one the remaining points! Hopefully we might be able to simplify/eliminate the workload you were taking on in #23947 to make up for it :)

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs:205]: Good feedback -- I've updated the comment to clarify the intent a little bit more.

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (captainsafia) [reply src/Components/Components/test/Routing/RouterTest.cs:60]: Good catch!

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (captainsafia) [reply src/Components/Components/test/Routing/RouterTest.cs:102]: Yeah, this is a cruft from a previous implementation of the tests where I was mocking the route table. Since we're mocking the `Refresh` method now we do not need it.

File: src/Components/Components/src/Routing/Router.cs
Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs:230]: > If the developer's task is cancelled by them, is the intended behavior that we treat it the same as if it resolved as success?

This is a correct. If the user cancels the task themselves, we'll still render the route's component. The only time we don't render the route's component is if the `OnNavigateAsync` was cancelled because another one was initiated. This effectively stops us from double rendering when a user quickly navigates between two routes that both have an `OnNavigateAsync`.

> The other option I'd consider is maybe throwing if the developer's own task was cancelled (and _onNavigateCts.Token was not). Because maybe that signals their intent to cancel navigation, and we don't support that. Throwing would leave us the option to add support for navigation cancellation in the future, whereas if we don't, then we couldn't use task cancellation as the signal for it.

It should be pretty easy to check this. We would have to do something like `if (completedTask.IsCancelled && !_onNavigateCts.Token.IsCancellationRequested) { throw InvalidOperationException(...); }`.

> But, how do we know it was cancelled by them?

We can do `task.IsCancelled` to validate this.

> Inside their OnNavigateAsync callback, they may well be using the cancellation token we gave them. So if that token fires, their returned task may be immediately cancelled as a result. If that happens it seems like we have a race condition: either task or cancellationTcs.Task might happen to resolve first, since they were both triggered by the same underlying _onNavigateCts.Token. It makes me wonder if L230 really should say return !_onNavigateCts.Token.IsCancellationRequested or something like that.

Maybe the more full-proof check is an adaptation of the one above `(completedTask.IsCancelled || _onNavigateCts.Token.IsCancellationRequested)` -- so if the task is cancelled by the user or via the token, then don't bother rendering the component's route.

> Also, if the developer's task resolves as faulted, then the logic here will disregard that fault and treat it the same as completing successfully. It seems like instead we should be throwing the exception upstream. I know that we don't yet have a way to truly pass it back into the renderer, but as it stands we're discarding the exception in two places (here on L230 and also on L264). Shouldn't we be throwing here so the fault goes to one place (L264) where we can do something useful with it later?

The whole "do something when a task throws an exception) is addressed in my other PR. You make a good point about the logic that throws the exception back to the renderer being the "OnNavigateAsyncWithRefresh" method instead of the `RunOnNavigateAsync` method.

File: src/Components/Components/test/Routing/RouterTest.cs
Comment (captainsafia) [reply src/Components/Components/test/Routing/RouterTest.cs:90]: > I'd love to see some additional test cases covering scenarios where the OnNavigateAsync task itself is what completes first, in success, failure, and cancelled status. 

These are good ideas. I'll add these scenarios as unit tests in the other PR.

File: src/Components/Components/src/Routing/Router.cs
Comment (halter73) [reply src/Components/Components/src/Routing/Router.cs:209]: I recommended this when reviewing the the CTS usage. I didn't want to have to think about the possible consequences of the meat of RunOnNavigateWithRefreshAsync running concurrently of itself, so in my mind this simplified things. I cannot point to any specific bug this prevents though.

Comment (captainsafia) [reply src/Components/Components/src/Routing/Router.cs:209]: @halter73 Thanks for posting this response. I also updated the commenting around here in a follow-up PR to make ti clearer.

**Pull request merged**
